### PR TITLE
PR: parallel_replicas_for_non_replicated_merge_tree in subqueries

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed parallel_replicas_for_non_replicated_merge_tree being ignored in subqueries for non-replicated tables

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 45a82d5be61e09f4320fed3c7982d05de0396b21 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73584/45a82d5be61e09f4320fed3c7982d05de0396b21/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/58d0272d0d3f839e6a3fd402378648fe1abeed18/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Planner/findParallelReplicasQuery.cpp [hunk 1]
```diff
@@ -38,12 +37,29 @@ namespace ErrorCodes
...
+    const auto & storage = table_node.getStorage();
+
+    if (!storage->isMergeTree() && !typeid_cast<const StorageDummy *>(storage.get()))
+        return false;
+
+    if (!storage->supportsReplication() && !context->getSettingsRef()[Setting::parallel_replicas_for_non_replicated_merge_tree])
+        return false;
+
+    /// Parallel replicas not supported with FINAL.
+    if (table_node.getTableExpressionModifiers() && table_node.getTableExpressionModifiers()->hasFinal())
```
Comment (nickitat) [inline src/Planner/findParallelReplicasQuery.cpp:51]:
```suggestion
    if (table_node.hasTableExpressionModifiers() && table_node.getTableExpressionModifiers()->hasFinal())
```

File: tests/queries/0_stateless/03290_pr_non_replicated_in_subquery.sql
[the remark this answers is not in the document]
Comment (devcrafter) [reply tests/queries/0_stateless/03290_pr_non_replicated_in_subquery.sql:1]: Rewrote the test and added a comment

**Review by nickitat — APPROVED** [review]

**Comment by devcrafter** [comment]
@nickitat It turns out that I've reverted the code from https://github.com/ClickHouse/ClickHouse/pull/63861 with commit https://github.com/ClickHouse/ClickHouse/pull/73584/commits/cf02794ffc399acd427879b76d61a27bd9d55131. The test `03143_parallel_replicas_mat_view_bug.sql` introduced in https://github.com/ClickHouse/ClickHouse/pull/63861 didn't fail, and local check shows that adding the check for MV back doesn't use PR while querying MV with enabled PR settings. Maybe you have some other example of query with MV where "we will end up executing a normal distributed subquery on the replicas, meaning each one will read the whole table without syncing anyhow with the initiator." (quote from https://github.com/ClickHouse/ClickHouse/pull/63861 description)

**Pull request merged**